### ISSUE TITLE
Otel Java agent proxy config

### ISSUE THREAD

**Issue by jpvallez** [open-telemetry/opentelemetry-java-instrumentation#12240]
Hi Otel community!

Much like https://github.com/open-telemetry/opentelemetry-java/issues/6204, I am trying to export from the otel java agent to an otlp endpoint through a proxy. We are unfortunately unable to use the collector in this instance so we’re looking for a way to configure proxy settings on the java agent.

I note that the feature to enable this has been added in https://github.com/open-telemetry/opentelemetry-java/pull/6270.

I’m not sure if it’s possible but - Can we configure proxy settings for the otlp exporter in the otel java agent by setting for example jvm flags or any other config? 
Or at this stage is it only possible by interacting with code/writing an extension as mentioned in https://github.com/open-telemetry/opentelemetry-java/issues/6204#issuecomment-2072541892

I also note the feature request here https://github.com/open-telemetry/opentelemetry-specification/issues/4007 which would be the most ideal solution to our problem.

Thanks!

**Comment by sxjthefirst** [comment]
We have this issue too. Sometimes we can't have a local collector in environments and have to traverse a proxy to get to one. 
We have two issues: 
1. JAVA Agent seems to not be able to connect to HTTP/S_PROXY set in env or with a -D equivalent and throws an error.
2. Ideally we want to use a proxy setting for ONLY the agent so that we don't affect any other connections the monitored java process might be making.

**Comment by breedx-splk** [comment]
Yeah right now it would need to be done with code, using the SPIs (see the [ConfigurableSpanExporterProvider](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/traces/ConfigurableSpanExporterProvider.java) and the config PR mentioned above).

At least before the spec gets exporter proxy settings, I think an agent extension that configures the exporter for proxy would be a worthy addition to [opentelemetry-java-contrib](https://github.com/open-telemetry/opentelemetry-java-contrib/), if you're willing to build and maintain such a thing.

**Comment by laurit** [comment]
The agent can be customized via [extensions](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension). Extensions can implement the SPIs from https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure-spi and use them to customize the opentelemetry sdk used by the agent. Unfortunately there doesn't seem to be an SPI that would let you easily configure the proxy settings. I guess the `ConfigurableSpanExporterProvider` mentioned by @breedx-splk or perhaps https://github.com/open-telemetry/opentelemetry-java/blob/18d192d1fadc6f6dd4dfbff01c0c139eaf93c604/sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/AutoConfigurationCustomizer.java#L64 could be used. You can use `addSpanExporterCustomizer` to replace the auto configured exporter with another one configured by yourself that has the proxy settings. This isn't really ideal as you will need to configure the exporter from scratch.
@jkwatson could this issue be moved back to the core repository so that @jack-berg can take a look. In my opinion this isn't as much about the agent but more about how to configure the proxy with autoconfigured sdk and whether something could be done to make such configuration less painful.

**Comment by jack-berg** [comment]
> This isn't really ideal as you will need to configure the exporter from scratch.

You can "enhance" the exporter by calling `toBuilder()` with something like:

```
public class ProxyCustomizer implements AutoConfigurationCustomizerProvider {
  @Override
  public void customize(AutoConfigurationCustomizer autoConfiguration) {
    autoConfiguration.addSpanExporterCustomizer((spanExporter, configProperties) -> {
      if (spanExporter instanceof OtlpHttpSpanExporter) {
        spanExporter.shutdown().join(10, TimeUnit.SECONDS);
        return ((OtlpHttpSpanExporter) spanExporter)
            .toBuilder()
            .setProxy(ProxyOptions.create(InetSocketAddress.createUnresolved("host", 8080)))
            .build();
      }
      return spanExporter;
    });
  }
}
```

In terms of making it less painful, [declarative configuration](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/configuration#declarative-configuration) is my answer. I would like to see proxy options be part of the OTLP exporter schema in the declarative configuration data model, which would avoid the need for any customization.

**Comment by sxjthefirst** [comment]
@jack-berg just to clarify do we need the declarative config update for creating a new otel parameter for the java agent?

I am not sure if that also includes the agent not being able to connect using the JVM level proxy. 

We get an error even if we start as below  (bug? agent doesn't read proxy?):
```
export HTTP_PROXY=blah:bleh
export HTTPS_PROXY=blah:bleh
java -Dotel.service.name="testapp" myapp
```

**Comment by trask** [comment]
@sxjthefirst can you try instead with system properties here https://docs.oracle.com/javase/8/docs/technotes/guides/net/proxies.html?

**Comment by jpvallez** [comment]
Hi everyone, thanks for the responses.

@trask - setting proxy with system properties as per https://docs.oracle.com/javase/8/docs/technotes/guides/net/proxies.html seems to work. For those following along, note this is a jvm wide setting.
E.g. `java -javaagent:opentelemetry-javaagent.jar -Dhttp.proxyHost=proxyhost -Dhttp.proxyPort=port -jar jar-file.jar`

This answers my question but still very keen on something exporter specific such as declarative configuration.

Thanks!

**Comment by Starefossen** [comment]
Same as the author but we set proxy and would like it to exempt the specific host like so where we excempt the the namespace the collector is running in from the proxy list:

```
- name: JAVA_PROXY_OPTIONS
          value: -Dhttp.proxyHost=some.proxy.org -Dhttps.proxyHost=some.proxy.org -Dhttp.proxyPort=8088 -Dhttps.proxyPort=8088 -Dhttp.nonProx
yHosts=*.some-namespace
```

**Comment by trask** [comment]
@Starefossen can you try setting `http.nonProxyHosts` (from https://docs.oracle.com/javase/8/docs/technotes/guides/net/proxies.html)?

I'm not certain whether OkHTTP honors this or not, but it's worth a try.

**Comment by Starefossen** [comment]
@trask `-Dhttp.nonProxyHosts=*.some-namespace` is set via `JAVA_PROXY_OPTIONS` and does not apear to be respected by OkHTTP in this instance.

**Comment by krismarc** [comment]
> @trask `-Dhttp.nonProxyHosts=*.some-namespace` is set via `JAVA_PROXY_OPTIONS` and does not apear to be respected by OkHTTP in this instance.

While java application has no [java proxies](https://docs.oracle.com/javase/8/docs/technotes/guides/net/proxies.html) configured then it works fine, I am able to connect. After adding java proxy settings, it starts to fail with timeouts.

> [otel.javaagent 2024-10-15 12:32:32:245 +0200] [OkHttp https://otel.ingest.***.com/...] WARN io.opentelemetry.exporter.internal.grpc.GrpcExporter - Failed to export spans. Server responded with gRPC status code 2. Error message: timeout

Therefore, I assume, it understands there's proxy configured like `-Dhttp.proxyHost` however, it doesn't honor `-Dhttp.nonProxyHosts` (or.. actual value syntax is misinterpreted? it contains pipes as separator). Please note our endpoint is considered internal, no proxy needed to reach it. While proxy is needed to reach things behind it for other application purposes. So no matter if I tried to whitelist our target within nonProxyHosts it tries to use proxy and we get a timeout.


Best regards,
K.M.
